pkg/util/admission/work_queue.go

```diff
@@ -456,8 +465,30 @@ type tenantHeap []*tenantInfo
 
 var _ heap.Interface = (*tenantHeap)(nil)
 
+var tenantInfoPool = sync.Pool{
+	New: func() interface{} {
+		return &tenantInfo{}
+	},
+}
+
 func makeTenantInfo(id uint64) *tenantInfo {
-	return &tenantInfo{id: id, heapIndex: -1}
+	ti := tenantInfoPool.Get().(*tenantInfo)
+	*ti = tenantInfo{
+		id:              id,
+		waitingWorkHeap: ti.waitingWorkHeap[:0],
+		heapIndex:       -1,
+	}
+	return ti
+}
+
+func releaseTenantInfo(ti *tenantInfo) {
+	if cap(ti.waitingWorkHeap) > 100 {
```

Comment:
Does this heap ever have non-nil values? If so, will it be sad to be keeping references to those objects alive? I notice you truncate in the `makeTenantInfo` which makes me think maybe.